### PR TITLE
Densenet with Python

### PR DESCRIPTION
**Pull request by forsyth2**
Set up DenseNet using new Python front end.

This pull request implements DenseNet, but we want to improve test accuracy - see #1055

### CODE REVIEW

**Comment by forsyth2** [comment]
@timmoon10 Ready for code review.

**Comment by ndryden** [comment]
Have we validated the accuracy?

File: model_zoo/vision/densenet_batch_job.cmd [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+#SBATCH --nodes 16
+#SBATCH --partition pbatch
+#SBATCH --time 240
+./densenet.py --nodes 16 --procs-per-node 2 --mini-batch-size 256 --num-epochs 10 > /usr/workspace/wsb/forsyth2/lbann/model_zoo/vision/output.txt
```
[the remark this answers is not in the document]
Comment (forsyth2) [reply model_zoo/vision/densenet_batch_job.cmd]: Ok, maybe I'll just add it as a block commented example in `densenet.py`.

Looks like `nodes` defaults to just 1 node in https://github.com/LLNL/lbann/blob/develop/python/lbann/launcher/slurm.py . Are `pbatch` and maximum time the defaults for `nodes` and `partition`?

**Comment by forsyth2** [comment]
@ndryden Test top 5% accuracy on 10 epochs is up to 77.788%. That's an error rate of 22.212% which seems close enough to the error rate of 16% suggested by https://arxiv.org/pdf/1608.06993.pdf page 7, Figure 4 (right).

File: model_zoo/vision/densenet_batch_job.cmd (hunk 1 above)
[the remark this answers is not in the document]
Comment (timmoon10) [reply model_zoo/vision/densenet_batch_job.cmd]: The default behavior of `lbann.run` is to not put anything for the partition or time limit, which I think defaults to pbatch and max time limit. With `lbann.contrib.lc.run`, the default is explicitly the pbatch partition on Pascal.

[the remark this answers is not in the document]
Comment (forsyth2) [reply model_zoo/vision/densenet_batch_job.cmd]: Ok sounds good. I'll make it a comment in the file so I have it for reference. Then it's good to merge?

File: model_zoo/vision/densenet.py [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+# --procs-per-node 2
+
+# To run the full 90 epochs from ssh:
+# ./densenet.py --nodes 16 --procs-per-node 2 > /usr/workspace/wsb/<username>/lbann/model_zoo/vision/output.txt
+# mini-batch-size default => 256, num-epochs => 90
+
+# To run 10 epoch test from ssh:
+# ./densenet.py --nodes 16 --procs-per-node 2 --mini-batch-size 256 --num-epochs 10 > /usr/workspace/wsb/<username>/lbann/model_zoo/vision/output.txt
+
+# To avoid needing to stay logged into ssh, create a script
```
Comment (forsyth2) [inline model_zoo/vision/densenet.py]: @timmoon10 Ok, made the batch script into an example comment.

**Comment by bvanessen** [comment]
I am not sure that we should say that we have a complete implementation if we are ~7% off.  Do we know why that is?  Are we doing the same data augmentation?

Brian Van Essen
vanessen1@llnl.gov<mailto:vanessen1@llnl.gov>
(w) 925-422-9300
(c) 925-290-5470

On May 21, 2019, at 6:09 PM, forsyth2 <notifications@github.com<mailto:notifications@github.com>> wrote:


@ndryden<https://github.com/ndryden> Test top 5% accuracy on 10 epochs is up to 77.788%. That's an error rate of 22.212% which seems close enough to the error rate of 16% suggested by https://arxiv.org/pdf/1608.06993.pdf page 7, Figure 4 (right).

**Pull request merged**